### DISCUSSION TITLE
convert value, unc, unit input to output in value, unit and relative unc

### DISCUSSION THREAD

**Discussion by bytelinker** [jagerber48/sciform#12] [Ideas]
sform = sciform.Formatter(precision=2, unicode_pm=True)
#input
a_value = 1.23456789
a_unc = 0.056789
#sform = sciform.Formatter(precision=2, unicode_pm=True, output_unc='percent'|'ppm', unit_str='unit)
a_unc_rel_percent = 100 * a_unc/a_value
#round a_value by precision and a_unc, calculate a_unc_rel_percent|ppm, round result by precision
print(f'unc_rel = {a_unc_rel_percent:.1f} %')
print(f'({sform(a_value, a_unc)}) unit')
print(f'{a_value:.3f} unit ± {4.6} %')
#output:
'''
unc_rel = 4.6 %
(1.235 ± 0.057) unit
1.235 unit ± 4.6 %
'''

**Comment by jagerber48** [comment]
@bytelinker I need more explanation here. Your example code is a confusing mix of things that will work with and without any changes to `sciform`. I think I see two or three requests:

- The ability to prepend a user-specified "unit" string onto the end of a formatted number. This is an idea I've tossed around, but it definitely starts moving more deeply into being a "units" formatting package rather than a just "number" formatting package. Something I'm not totally sure I want to do. The user can always easily add a simple unit string on themselves.
- The second request involves a "relative" uncertainty presentation where (1) math is done on the absolute uncertainty and the value to determine the relative uncertainty and then (2) the relative uncertainty is presented as a percentage or ppm from the value. E.g. 1.00 +/- 0.01 would be presented as (1.00 +/- 1%) and 1.00000 +/- 0.00001 would be presented as (1.00000 +/- 10 ppm)
- Finally, combining the previous two, the ability to present relative uncertainty WITH units. E.g. 1.00 kHz +/- 1%. For this one, since the unit string is embedded in the  string, it is necessary for the unit string functionality to be a part of `sciform`, i.e., users can't do this themselves after the fact.

One multipart question:
- For the relative uncertainty: How should the number of sig figs to show for the value and uncertainty be determined? Should the sig figs for the value be determined as if the uncertainty was shown in absolute mode? How should the sig figs for the uncertainty be determined? Can you provide links (and instructions to specific sections) to specific NIST/BIPM recommendations on formatting with relative uncertainty? I haven't seen relative uncertainty formatting much, if at all. I'd like to also see recommendations that show the unit string should be "embedded" as you've requested for relative uncertainty.

**Comment by bytelinker** [comment]
- To 1: Adding one feature can cause a lot of additional feature requests. That's true.
- To 2: difficult to use, as the unit of the first parts is difficult to "inject"
- To 3: True
- To 4: I calculated manually in the example. I can work out an example. There are comments on % and ppm in the NIST documentation.

**Reply by jagerber48** [reply]
I'd be curious to see details about your calculations for how to determine # of sig figs for value and uncertainty when displaying relative uncertainty. I'd also appreciate if you could point me to a specific section in a specific NIST document to find information about % and ppm. I haven't seen information about relative uncertainty yet, but I certainly haven't combed the documents looking for it yet either.

**Reply by jagerber48** [reply]
The only guidance from NIST I've found is here: https://www.nist.gov/pml/special-publication-811/nist-guide-si-chapter-7-rules-and-style-conventions-expressing-values sec 7.10.2 and 7.10.3.

7.10.2 indicates that use of the % symbol is appropriate, but gives no guidance about formatting values together with relative uncertainty expressed as a percentage.

7.10.3 generally discourages all use of "ppm" type notation. Largely names like "billion" do not correspond to the same order of magnitude throughout different locales and NIST discourages use of language dependent and independent nomenclatures.

**Comment by bytelinker** [comment]
There might be more examples like: https://deanza.edu/faculty/lunaeduardo/documents/UncertaintyandSignificantFig.pdf page 5/5. There are more examples in tables, but the value und unc are written in separate columns. Case 7.10.3 could be covered by a strict option, which is true by default and disables 7.10.3, so the used could enable it.

**Reply by jagerber48** [reply]
The linked example is 47 +/- 2. 2/47 = 0.04255319148... I'm not sure if the decimal representation is finite or infinite. The author of that pdf truncates this to 0.043. Why? Why not 0.04 so that we have 47 +/- 4%? It would help if you could point me towards official documentation for this type of formatting. There are two feature requests here:

(1) support for a presently partially specified percent format relative uncertainty and
(2) support for embedding units within this relative uncertainty format.

Regarding (1), it's not clear that this formatting is supported by official bodies like NIST or BIPM so I'm hesitant to include anything. Does siunitx for latex formatting include this sort of relative formatting? If so could you point me to it in the documentation?

Regarding (2): It's true that units are an important part of scientific formatting, but so far I've been happy to keep units OUT of sciform because proper handling of units can itself be a very complex and challenging subject. Of course a first pass could just use a user supplied string as the "units". But again, the only use case so far where the user can't just append units to the end of the string provided by `sciform` is case (1) that you've brought up here, and again, it doesn't seem like that format is officially recommended. 

In any case, I would address (1) first and then reconsider (2). But to consider (1) I need to see instruction for how to determine the number of sig figs on the relative uncertainty, so that would be the next step.
